core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -3883,6 +3883,59 @@ private static AtomicLong getAtomicLong(String key) {
     return atomic;
   }
 
+  /** Support the ARRAYS_OVERLAP function. */
+  public static @Nullable Boolean arraysOverlap(List list1, List list2) {
+    final List bigger = list1.size() > list2.size() ? list1 : list2;
+    final List smaller = list1.size() > list2.size() ? list2 : list1;
+    boolean hasNull = false;
```

Comment:
could this be simplified using `.retainsAll()` or some other set operation like an `intersect`? I guess the `NULL` behavior could make that more difficult, just wondering.

Reply:
`retainAll` won't work because it mutates one list.

However, we have `Util.intersects(Collection, Collection)`. It's even tested. However it assumes that there are no null values.

Maybe you could quickly check for nulls and then call `Util.intersects` when you've determined that there are no nulls.

Reply:
hi @julianhyde @tanclary 
1) why Util.intersects is fast, it has O(n * m) time space, and my implementation just has O(n+m) time space.
```
  public static <E> boolean intersects(Collection<E> c0, Collection<E> c1) {
    for (E e : c1) {
      if (c0.contains(e)) {
        return true;
      }
    }
    return false;
  }
```

2)  about null
Returns true if *array1 contains at least a non-null element present also in *array2*. If the arrays have no common element and they are both non-empty and either of them contains a null element null is returned, false otherwise
  
if you look the three unit test, you will know why, and the implementation is align with spark
f.checkScalar("arrays_overlap(array[1, null], array[1])", true,
        "BOOLEAN");
f.checkNull("arrays_overlap(array[1], array[2, null])");
f.checkScalar("arrays_overlap(array[1, 2], array[3])", false,
        "BOOLEAN NOT NULL");

Reply:
Yes, your algorithm is probably more efficient.

Note that `intersects` uses `contains`, which is possibly optimized using intrinsics.

You could assign smaller and larger using one test rather than two. In fact, to keep it simple, I'd probably write
```
if (list1.size() > list2.size()) {
  return arraysOverlap(list2, list1);
}
```

Can you add a unit test to `SqlFunctionsTest`? There are at least 16 combinations to check: each list can be empty, contain only nulls, contain a mixture of nulls and non-nulls, and contain only non-nulls, which gives 4 * 4. That's easier done in a Java unit test.

Reply:
Also,
```
final Set smallestSet = new HashSet(smaller);
hasNull = smallestSet.remove(null);
```
is probably more efficient because it allows the set to right-size the table.

Reply:
hi @julianhyde added the unit test, thanks!